File: sidecar/src/sidecar/domain/model_readiness.py
"""How far the retrieval model is from being usable.

First run fetches about four and a half gigabytes before anything can be
searched, and a progress bar is the difference between waiting and thinking
the app is broken. What the states mean lives here; who reports them is the
embedder's business.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum


class ModelState(StrEnum):
    ABSENT = "absent"
    DOWNLOADING = "downloading"
    LOADING = "loading"
    READY = "ready"


@dataclass(frozen=True)
class ModelProgress:
    """Where the model is now, and how much of the download is done.

    The byte counts are zero unless a download is running, because the only
    thing they are for is the bar drawn during one.
    """

    state: ModelState = ModelState.ABSENT
    bytes_done: int = 0
    bytes_total: int = 0

    @property
    def fraction(self) -> float | None:
        """How far along, or `None` while the total is still unknown.

        `None` rather than zero: a bar sitting at zero percent says the
        download is stuck, and at the start of one the total genuinely is not
        known yet. A caller shows a spinner until this is a number.
        """
        if self.state is not ModelState.DOWNLOADING or self.bytes_total <= 0:
            return None
        return min(1.0, self.bytes_done / self.bytes_total)

    @property
    def usable(self) -> bool:
        return self.state is ModelState.READY

File: app/src/renderer/domain/model-readiness.ts
export type ModelState = 'absent' | 'downloading' | 'loading' | 'ready'

export interface ModelReadiness {
  state: ModelState
  bytesDone: number
  bytesTotal: number
  /** Null while the total is unknown, so the UI shows a spinner rather than a bar stuck at zero. */
  fraction: number | null
}

export const modelUnknown: ModelReadiness = {
  state: 'absent',
  bytesDone: 0,
  bytesTotal: 0,
  fraction: null,
}

/**
 * What to say while the model is being made ready, or null when there is nothing to say.
 *
 * Silent when the model is ready and when it is merely absent: absent is the
 * resting state of a machine nobody has searched on yet, and announcing it
 * would put a warning on an app that is working correctly.
 */
export function describeModel(readiness: ModelReadiness): string | null {
  if (readiness.state === 'loading') return 'Starting the search model.'
  if (readiness.state !== 'downloading') return null
  if (readiness.bytesTotal <= 0) return 'Getting the search model. This happens once.'
  return `Getting the search model. ${gigabytes(readiness.bytesDone)} of ${gigabytes(readiness.bytesTotal)} GB.`
}

function gigabytes(bytes: number): string {
  return (bytes / 1024 ** 3).toFixed(1)
}
